### PR TITLE
Add BigBlueButton monitoring with a Node.js plugin

### PR DESCRIPTION
**Pull request by madmatah**
##### Summary

Add a Node.js plugin to monitor [BigBlueButton](https://bigbluebutton.org/).

##### Component Name

`node.d.plugin/bigbluebutton`

##### Test Plan

- I installed this plugin on my own netdata instance. BigBlueButton was running on the same server.
- I followed the instructions documented in the [README.md](https://github.com/madmatah/netdata/blob/98a6f7747258674fe23e8011fc66b83c2843d265/collectors/node.d.plugin/bigbluebutton/README.md) (add configuration file + set `url` and `secret`  options)

##### Additional Information

[Link to BigBlueButton API documentation](https://docs.bigbluebutton.org/dev/api.html).

The following charts and dimensions are added (see screenshot below) :

###### bbb_conferences
- **conferences** : the number of active conferences running on the server.

###### bbb_users
- **participants** : the total number of participants connected to a conference on the BigBlueButton server
- **listeners** : the number of participants connected in «Listen only» mode
- **voiceParticipants** : the number of participants connected in «Microphone» mode
- **videos** : the number of active video streams (webcam + screen sharing)

Here is a demo screenshot : 

bbb-netdata-plugin

This is my first contribution to the netdata project, so I'm interested by your feedbacks and help to improve this plugin.

### CODE REVIEW

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- collectors/node.d.plugin/bigbluebutton/bigbluebutton.node.js  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5989897&bid=19641944)

**Comment by Ferroin** [comment]
The LGTM security alert can be ignored, it's due to the requirements of the AP the module is interacting with.

File: collectors/node.d.plugin/bigbluebutton/bigbluebutton.node.js [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+      service.begin(chart);
+      service.set("conferences", conferences);
+      service.end();
+
+      var sumXmlAttr = function(name) {
+        var sum = 0;
+        var r = new RegExp("<" + name + ">([0-9]+)</" + name + ">", "g");
+        var match = r.exec(data);
+        while (match !== null) {
+          sum += parseInt(match[1]);
```
Comment (Ferroin) [inline collectors/node.d.plugin/bigbluebutton/bigbluebutton.node.js:59]:
```suggestion
          sum += parseInt(match[1], 10);
```

Pointed out by Codacy. Without the radix being explicitly specified there's a risk of this not actually decoding the integer as base 10.

**Comment by ilyam8** [comment]
i would say it is `wontfix`, we dont want one more collector in `nodejs`. I checked the code, it is http req + xml parse, i suggest to write in [go](https://github.com/netdata/go.d.plugin/tree/master/modules)

**Comment by ilyam8** [comment]
@madmatah we dont really have people to be nodejs stuff maintainers.

**Comment by Ferroin** [comment]
@ilyam8 I can't really cover proper maintainership for this stuff, but I've dealt with JS enough that I can at least provide code reviews for community submitted Node.JS collectors.

**Comment by ilyam8** [comment]
Well, i can only repeat, it is much better to not add new node.js collectors. This one is super easy to write in go. Even python is much better then nodejs.

File: collectors/node.d.plugin/bigbluebutton/bigbluebutton.node.js (hunk 1 above)
Comment (madmatah) [reply collectors/node.d.plugin/bigbluebutton/bigbluebutton.node.js]: @Ferroin Thanks for your review, I addressed your feedbacks.

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- collectors/node.d.plugin/bigbluebutton/bigbluebutton.node.js  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5989897&bid=19641944)

**Comment by madmatah** [comment]
> i would say it is `wontfix`, we dont want one more collector in `nodejs`. I checked the code, it is http req + xml parse, i suggest to write in [go](https://github.com/netdata/go.d.plugin/tree/master/modules)

@ilyam8  Is this documented somewhere ? I couldn't find anything in the documentation about the node.js plugins maintenance problem. I've only found one comment about go vs python plugins.

I suggest to explain that in the [collector QUICKSTART documentation](https://github.com/netdata/netdata/blob/master/collectors/QUICKSTART.md) or in the contributing guide.

**Comment by Ferroin** [comment]
> I suggest to explain that in the collector QUICKSTART documentation or in the contributing guide.

Seconded, if we're not accepting new Node.js plugins due to maintenance issues, we should document that. It may even result in someone from the community stepping up to maintain it.

**Comment by ilyam8** [comment]
We adding new features like service discovery to `go` plugin and we plan to continue do it. We addeded file lock registry to py and go plugin exactly for the migration, it helps us to start to do it more agressivly.

**We want `bash`, `nodejs` and `py` plugins to be only for user custom colletors (we have third-party plugin page btw).** 

Adding something to `nodejs` just makes migration complex, especially if there is no some real need (example - we need some library and there is no go native library and we dont want cgo).

This specific plugin is extremly easy to write in go. I said even in `py` is better because we have this `py` - `go` migration path (file lock registry).

It is not documented - yes, this should be fixed - yes. But at the same time this PR could start from an issue where we would discuss it.

@madmatah dont mind me wrong, i appreciate your contribution and this is very nice 👍, my complain is that this just add more problems with the migration.

**Comment by madmatah** [comment]
@ilyam8 Thank you for this explanation, I understand your concerns better.

I didn't start with an issue because the work has already been done for an internal usage in my company. We just wanted to share it.

I see 2 options here :
- I can rewrite it in python if there is a chance to be merged on the main repository.
- Otherwise, we will publish it as is as a third party plugin.

Since I don't know `go` and don't have time to learn now, i do not plan to rewrite it as a `go` plugin.
But if someone wants to rewrite it in go, I can help to test it !

**Comment by amoss** [comment]
@madmatah Thank you for your understanding and we appreciate your contribution. If you could rewrite it in Python then I think that we can merge it without any problems and then @ilyam8 can pick it up later for migration from Python to Go.

@ilyam8 / @Ferroin Could you update the contributors guide as suggested. I agree it would be good for us to document our future direction.

**Comment by ilyam8** [comment]
@madmatah rewriting in `py` is ok

File: collectors/node.d.plugin/bigbluebutton/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+title: "BigBlueButton monitoring with Netdata"
+custom_edit_url: https://github.com/netdata/netdata/edit/master/collectors/node.d.plugin/bigbluebutton/README.md
+sidebar_label: "BigBlueButton"
+-->
+
+# BigBlueButton monitoring with Netdata
+
+Example Netdata configuration for node.d/bigbluebutton.conf
+
+The module supports any number of servers, like this.
```
Comment (DShreve2) [inline collectors/node.d.plugin/bigbluebutton/README.md:11]:
```suggestion
The module supports any number of servers, like the following.
```

File: collectors/node.d.plugin/bigbluebutton/README.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+Auto detection is not supported.
```
Comment (DShreve2) [inline collectors/node.d.plugin/bigbluebutton/README.md:12]:
```suggestion
> **Note**: Auto detection is not supported.
```

**Review by DShreve2 — CHANGES_REQUESTED** [review]
Minor doc changes requested.

**Comment by ilyam8** [comment]
We have deprecated (removed) node.d plugin and all nodejs collectors.

**Pull request closed without merging**